File: lib/services/coin_base.py
import logging
from datetime import datetime
from coinbase.wallet.client import Client
from lib.services.base import BaseService
from lib import configs
from lib.services.currency_convert_service import CurrencyConvertService


def client():
    return Client(configs.COINBASE_API_KEY, configs.COINBASE_API_SECRET)


class CoinBaseService(BaseService):
    EXCHANGE_NAME = 'coinbase'

    def __init__(self):
        self.__client = Client(configs.COINBASE_API_KEY, configs.COINBASE_API_SECRET)
        self.__currency_converter = CurrencyConvertService()

    def get_balances(self):
        logging.info('Start query for CoinBase balance')
        now = datetime.utcnow()
        balances_data = []
        accounts = self.__client.get_accounts().data
        for account in accounts:
            logging.info('Current CoinBase %s balance: %s %s' % (account['name'],
                                                        account['native_balance']['currency'],
                                                        account['native_balance']['amount']))
            sgd_amount = float(account['native_balance']['amount'])
            balance_data = {
                'timestamp': now,
                'exchange': self.EXCHANGE_NAME,
                'type': account['type'],
                'account': account['name'],
                'currency': account['currency'],
                'balances': {
                    'native': float(account['balance']['amount']),
                    'sgd': sgd_amount,
                    'usd': self.__currency_converter.convert(sgd_amount, 'SGD', 'USD')
                }
            }
            balances_data.append(balance_data)
        return balances_data

File: internal/services/cryto_exchanges/coinbase/coinbase.go
package coinbase

import (
	"fmt"
	"github.com/Zauberstuhl/go-coinbase"
	"github.com/seannguyen/coin-tracker/internal/services/cryto_exchanges"
	"github.com/spf13/viper"
)

const exchangeName = "coinbase"

type Exchange struct{}

func (*Exchange) GetBalances() ([]*cryto_exchanges.BalanceData, error) {
	client := getApiClient()
	return getBalances(&client)
}

func getApiClient() coinbase.APIClient {
	return coinbase.APIClient{
		Key:    viper.GetString("COINBASE_API_KEY"),
		Secret: viper.GetString("COINBASE_API_SECRET"),
	}
}

func getBalances(client *coinbase.APIClient) ([]*cryto_exchanges.BalanceData, error) {
	accounts, err := client.Accounts()
	if err != nil {
		fmt.Println(err)
		return nil, nil
	}

	balanceByCurrency := make(map[string]float64)
	for _, accountData := range accounts.Data {
		balanceByCurrency[accountData.Currency] += accountData.Balance.Amount
	}

	balances := make([]*cryto_exchanges.BalanceData, 0)
	for currency, balance := range balanceByCurrency {
		if balance <= 0 {
			continue
		}
		balanceData := cryto_exchanges.BalanceData{
			Currency:     currency,
			Amount:       balance,
			ExchangeName: exchangeName,
			Type:         cryto_exchanges.Crypto,
		}
		balances = append(balances, &balanceData)
	}

	return balances, nil
}
